### PR TITLE
ModelChainResult.__repr__

### LINKED ISSUE #1110: Improve ModelChainResult.__repr__
did not require in #1076... we should improve the `ModelChainResult.__repr__`. See details below for what it currently looks like. A simple improvement might be a new line between the attribute and data, then a blank line between the end of the data and the next attribute name. Perhaps some indentation would help keep track of results from multiple arrays.

<details>

```
ModelChainResult(solar_position=                           apparent_zenith      zenith  apparent_elevation  \
2020-12-01 00:00:00-07:00       169.300758  169.300758          -79.300758   
2020-12-01 00:05:00-07:00       169.539192  169.539192          -79.539192   
2020-12-01 00:10:00-07:00       169.663237  169.663237          -79.663237   
2020-12-01 00:15:00-07:00       169.668762  169.668762          -79.668762   
2020-12-01 00:20:00-07:00       169.555581  169.555581          -79.555581   
...                                    ...         ...                 ...   
2020-12-02 23:40:00-07:00       167.518383  167.518383          -77.518383   
2020-12-02 23:45:00-07:00       168.144707  168.144707          -78.144707   
2020-12-02 23:50:00-07:00       168.696851  168.696851          -78.696851   
2020-12-02 23:55:00-07:00       169.163414  169.163414          -79.163414   
2020-12-03 00:00:00-07:00       169.532897  169.532897          -79.532897   

                           elevation     azimuth  equation_of_time  
2020-12-01 00:00:00-07:00 -79.300758  343.920897         10.897119  
2020-12-01 00:05:00-07:00 -79.539192  350.104880         10.895814  
2020-12-01 00:10:00-07:00 -79.663237  356.495783         10.894508  
2020-12-01 00:15:00-07:00 -79.668762    2.963292         10.893202  
2020-12-01 00:20:00-07:00 -79.555581    9.367007         10.891897  
...                              ...         ...               ...  
2020-12-02 23:40:00-07:00 -77.518383  321.421138         10.129835  
2020-12-02 23:45:00-07:00 -78.144707  326.040347         10.128458  
2020-12-02 23:50:00-07:00 -78.696851  331.110143         10.127082  
2020-12-02 23:55:00-07:00 -79.163414  336.628915         10.125705  
2020-12-03 00:00:00-07:00 -79.532897  342.565345         10.124328  

[577 rows x 6 columns], airmass=                           airmass_relative  airmass_absolute
2020-12-01 00:00:00-07:00               NaN               NaN
2020-12-01 00:05:00-07:00               NaN               NaN
2020-12-01 00:10:00-07:00               NaN               NaN
2020-12-01 00:15:00-07:00               NaN               NaN
2020-12-01 00:20:00-07:00               NaN               NaN
...                                     ...               ...
2020-12-02 23:40:00-07:00               NaN               NaN
2020-12-02 23:45:00-07:00               NaN               NaN
2020-12-02 23:50:00-07:00               NaN               NaN
2020-12-02 23:55:00-07:00               NaN               NaN
2020-12-03 00:00:00-07:00               NaN               NaN

[577 rows x 2 columns], ac=2020-12-01 00:00:00-07:00    0.0
2020-12-01 00:05:00-07:00    0.0
2020-12-01 00:10:00-07:00    0.0
2020-12-01 00:15:00-07:00    0.0
2020-12-01 00:20:00-07:00    0.0
                            ... 
2020-12-02 23:40:00-07:00    0.0
2020-12-02 23:45:00-07:00    0.0
2020-12-02 23:50:00-07:00    0.0
2020-12-02 23:55:00-07:00    0.0
2020-12-03 00:00:00-07:00    0.0
Freq: 5T, Length: 577, dtype: float64, total_irrad=                           poa_global  poa_direct  poa_diffuse  \
2020-12-01 00:00:00-07:00         0.0        -0.0          0.0   
2020-12-01 00:05:00-07:00         0.0        -0.0          0.0   
2020-12-01 00:10:00-07:00         0.0        -0.0          0.0   
2020-12-01 00:15:00-07:00         0.0        -0.0          0.0   
2020-12-01 00:20:00-07:00         0.0        -0.0          0.0   
...                               ...         ...          ...   
2020-12-02 23:40:00-07:00         0.0        -0.0          0.0   
2020-12-02 23:45:00-07:00         0.0        -0.0          0.0   
2020-12-02 23:50:00-07:00         0.0        -0.0          0.0   
2020-12-02 23:55:00-07:00         0.0        -0.0          0.0   
2020-12-03 00:00:00-07:00         0.0        -0.0          0.0   

                           poa_sky_diffuse  poa_ground_diffuse  
2020-12-01 00:00:00-07:00              0.0                 0.0  
2020-12-01 00:05:00-07:00              0.0                 0.0  
2020-12-01 00:10:00-07:00              0.0                 0.0  
2020-12-01 00:15:00-07:00              0.0                 0.0  
2020-12-01 00:20:00-07:00              0.0                 0.0  
...                                    ...                 ...  
2020-12-02 23:40:00-07:00              0.0                 0.0  
2020-12-02 23:45:00-07:00              0.0                 0.0  
2020-12-02 23:50:00-07:00              0.0                 0.0  
2020-12-02 23:55:00-07:00              0.0                 0.0  
2020-12-03 00:00:00-07:00              0.0                 0.0  

[577 rows x 5 columns], aoi=2020-12-01 00:00:00-07:00    169.300758
2020-12-01 00:05:00-07:00    169.539192
2020-12-01 00:10:00-07:00    169.663237
2020-12-01 00:15:00-07:00    169.668762
2020-12-01 00:20:00-07:00    169.555581
                                ...    
2020-12-02 23:40:00-07:00    167.518383
2020-12-02 23:45:00-07:00    168.144707
2020-12-02 23:50:00-07:00    168.696851
2020-12-02 23:55:00-07:00    169.163414
2020-12-03 00:00:00-07:00    169.532897
Freq: 5T, Name: aoi, Length: 577, dtype: float64, aoi_modifier=2020-12-01 00:00:00-07:00    0.0
2020-12-01 00:05:00-07:00    0.0
2020-12-01 00:10:00-07:00    0.0
2020-12-01 00:15:00-07:00    0.0
2020-12-01 00:20:00-07:00    0.0
                            ... 
2020-12-02 23:40:00-07:00    0.0
2020-12-02 23:45:00-07:00    0.0
2020-12-02 23:50:00-07:00    0.0
2020-12-02 23:55:00-07:00    0.0
2020-12-03 00:00:00-07:00    0.0
Freq: 5T, Length: 577, dtype: float64, spectral_modifier=1, cell_temperature=2020-12-01 00:00:00-07:00    20.0
2020-12-01 00:05:00-07:00    20.0
2020-12-01 00:10:00-07:00    20.0
2020-12-01 00:15:00-07:00    20.0
2020-12-01 00:20:00-07:00    20.0
                             ... 
2020-12-02 23:40:00-07:00    20.0
2020-12-02 23:45:00-07:00    20.0
2020-12-02 23:50:00-07:00    20.0
2020-12-02 23:55:00-07:00    20.0
2020-12-03 00:00:00-07:00    20.0
Freq: 5T, Length: 577, dtype: float64, effective_irradiance=2020-12-01 00:00:00-07:00    0.0
2020-12-01 00:05:00-07:00    0.0
2020-12-01 00:10:00-07:00    0.0
2020-12-01 00:15:00-07:00    0.0
2020-12-01 00:20:00-07:00    0.0
                            ... 
2020-12-02 23:40:00-07:00    0.0
2020-12-02 23:45:00-07:00    0.0
2020-12-02 23:50:00-07:00    0.0
2020-12-02 23:55:00-07:00    0.0
2020-12-03 00:00:00-07:00    0.0
Freq: 5T, Length: 577, dtype: float64, dc=2020-12-01 00:00:00-07:00    0.0
2020-12-01 00:05:00-07:00    0.0
2020-12-01 00:10:00-07:00    0.0
2020-12-01 00:15:00-07:00    0.0
2020-12-01 00:20:00-07:00    0.0
                            ... 
2020-12-02 23:40:00-07:00    0.0
2020-12-02 23:45:00-07:00    0.0
2020-12-02 23:50:00-07:00    0.0
2020-12-02 23:55:00-07:00    0.0
2020-12-03 00:00:00-07:00    0.0
Freq: 5T, Length: 577, dtype: float64, array_ac=None, diode_params=None)
```

</details>

_Originally posted by @wholmgren in https://github.com/pvlib/pvlib-python/issues/1076#issuecomment-741882154_

### PR DESCRIPTION
**Pull request by cwhanse**
- [x] Closes #1110
 - [x] I am familiar with the [contributing guidelines](https://pvlib-python.readthedocs.io/en/latest/contributing.html)
 - ~~[ ] Tests added~~
 - [x] Updates entries to [`docs/sphinx/source/api.rst`](https://github.com/pvlib/pvlib-python/blob/master/docs/sphinx/source/api.rst) for API changes.
 - ~~[ ] Adds description and name entries in the appropriate "what's new" file in [`docs/sphinx/source/whatsnew`](https://github.com/pvlib/pvlib-python/tree/master/docs/sphinx/source/whatsnew) for all changes. Includes link to the GitHub Issue with `` :issue:`num` `` or this Pull Request with `` :pull:`num` ``. Includes contributor name and/or GitHub username (link with `` :ghuser:`user` ``).~~
 - [x] New code is fully documented. Includes [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) compliant docstrings, examples, and comments where necessary.
 - [x] Pull request is nearly complete and ready for detailed review.
 - [x] Maintainer: Appropriate GitHub Labels and Milestone are assigned to the Pull Request and linked Issue.

### CODE REVIEW

**Comment by cwhanse** [comment]
I'm not unhappy with the print of ModelChainResult here. The code is clumsy, because ModelChainResult doesn't hew to any standards for the type of each attribute (e.g., with 2 Arrays, the value can be None, a single float, or a tuple of something).

Here's what the report looks like for a simple ModelChain (system with one Array). I'm "head"ing any Series or DataFrames.

Any if you read this far, look at the name of the last Series for `ModelChainResult.ac`.

```
=== ModelChainResult === 
weather 
                                   ghi         dhi  ...  temp_air  wind_speed
2023-06-10 10:00:00-07:00  775.999173  108.933862  ...        25           0
2023-06-10 11:00:00-07:00  898.985531  116.633882  ...        25           0
2023-06-10 12:00:00-07:00  969.592761  120.879109  ...        25           0
2023-06-10 13:00:00-07:00  982.786788  121.660833  ...        25           0
2023-06-10 14:00:00-07:00  937.637996  118.971374  ...        25           0

[5 rows x 5 columns] 

 solar_position 
                            apparent_zenith  ...  equation_of_time
2023-06-10 10:00:00-07:00        37.382226  ...          0.539325
2023-06-10 11:00:00-07:00        25.474558  ...          0.530968
2023-06-10 12:00:00-07:00        15.254955  ...          0.522605
2023-06-10 13:00:00-07:00        12.488151  ...          0.514235
2023-06-10 14:00:00-07:00        20.492312  ...          0.505857

[5 rows x 6 columns] 

 airmass 
                            airmass_relative  airmass_absolute
2023-06-10 10:00:00-07:00          1.257469          1.257469
2023-06-10 11:00:00-07:00          1.107109          1.107109
2023-06-10 12:00:00-07:00          1.036112          1.036112
2023-06-10 13:00:00-07:00          1.023853          1.023853
2023-06-10 14:00:00-07:00          1.067070          1.067070 

 
 Number of Arrays: 1 
------------------- 
  Array 0 
------------------- 
  tracking 
 None 

  aoi 
 2023-06-10 10:00:00-07:00    37.382226
2023-06-10 11:00:00-07:00    25.474558
2023-06-10 12:00:00-07:00    15.254955
2023-06-10 13:00:00-07:00    12.488151
2023-06-10 14:00:00-07:00    20.492312
Freq: H, Name: aoi, dtype: float64 

  aoi_modifier 
 1.0 

  total_irrad 
                            poa_global  ...  poa_ground_diffuse
2023-06-10 10:00:00-07:00  775.994182  ...                 0.0
2023-06-10 11:00:00-07:00  898.983256  ...                 0.0
2023-06-10 12:00:00-07:00  969.591954  ...                 0.0
2023-06-10 13:00:00-07:00  982.786249  ...                 0.0
2023-06-10 14:00:00-07:00  937.636532  ...                 0.0

[5 rows x 5 columns] 

  spectral_modifier 
 1 

  effective_irradiance 
 2023-06-10 10:00:00-07:00    775.994182
2023-06-10 11:00:00-07:00    898.983256
2023-06-10 12:00:00-07:00    969.591954
2023-06-10 13:00:00-07:00    982.786249
2023-06-10 14:00:00-07:00    937.636532
Freq: H, dtype: float64 

  cell_temperature 
 2023-06-10 10:00:00-07:00    51.474617
2023-06-10 11:00:00-07:00    55.670639
2023-06-10 12:00:00-07:00    58.079598
2023-06-10 13:00:00-07:00    58.529749
2023-06-10 14:00:00-07:00    56.989374
Freq: H, dtype: float64 

  dc 
 2023-06-10 10:00:00-07:00    173.454396
2023-06-10 11:00:00-07:00    197.173423
2023-06-10 12:00:00-07:00    210.324276
2023-06-10 13:00:00-07:00    212.743986
2023-06-10 14:00:00-07:00    204.414727
Freq: H, Name: p_mp, dtype: float64 

  dc_ohmic_losses 
 None 

 End of Arrays 
------------------- 

    losses 
1 
 
  ac 
2023-06-10 10:00:00-07:00    166.923393
2023-06-10 11:00:00-07:00    189.648371
2023-06-10 12:00:00-07:00    202.216782
2023-06-10 13:00:00-07:00    204.526893
2023-06-10 14:00:00-07:00    196.571725
Freq: H, Name: p_mp, dtype: float64 
```

**Comment by cwhanse** [comment]
@kandersolar your call if this is for v0.10 or later.

**Comment by kandersolar** [comment]
I think using `head` for each attribute takes up a ton of space in the output just to display what is going to be mostly boring nighttime values for most simulations.  How about just showing the type and shape of the data, but not any actual values?  For example, something like this:

```
ModelChainResult:
  weather: DataFrame (145 rows x 5 columns)
  solar_position: DataFrame (145 rows x 6 columns)
  airmass: DataFrame (145 rows x 2 columns)
  tracking: None
  aoi: Tuple (Series (length 145), Series (length 145))
  aoi_modifier: Tuple (Series (length 145), Series (length 145))
  total_irrad: Tuple (DataFrame (145 rows x 5 columns), DataFrame (145 rows x 5 columns))
  spectral_modifier: Tuple (1, 1)
  effective_irradiance: Tuple (Series (length 145), Series (length 145))
  cell_temperature: Tuple (Series (length 145), Series (length 145))
  dc: Tuple (Series (length 145), Series (length 145))
  dc_ohmic_losses: None
  losses: 1
  ac: Series (length 145)
```

Code for the above:
```python
def _mcr_repr(obj):
    if isinstance(obj, tuple):
        return "Tuple (" + ", ".join([_mcr_repr(o) for o in obj]) + ")"
    if isinstance(obj, pd.DataFrame):
        return "DataFrame ({} rows x {} columns)".format(*obj.shape)
    if isinstance(obj, pd.Series):
        return "Series (length {})".format(len(obj))
    # scalar, None, other?
    return repr(obj)

def __repr__(self):
    lines = ['ModelChainResult:']
    attrs = [
        'weather', 'solar_position', 'airmass',
        'tracking', 'aoi', 'aoi_modifier', 'total_irrad', 'spectral_modifier',
        'effective_irradiance', 'cell_temperature', 'dc', 'dc_ohmic_losses', 'losses', 'ac'
    ]
    for attr in attrs:
        lines.append(f"  {attr}: " + _mcr_repr(getattr(self, attr)))

    return "\n".join(lines)
```

**Comment by cwhanse** [comment]
I had three outcomes in mind:
- report number of Arrays
- report attributes and their shape. I was doing this by writing a report for each Array, but beyond 2 or 3 Arrays, that will be text barf on the screen.
- provide a summary of key data, e.g., the DatetimeIndex. I could do that once rather than using `head()` for each DataFrame or Series.
So I'll drop the `head` except for `times`

**Comment by cwhanse** [comment]
Code I'm using to display the report for different PVSystems

```
import pytz
import pandas as pd
import pvlib


loc = pvlib.location.Location(35, -116)
dr = pd.date_range(start='2023-06-10 10:00:00', end='2023-06-11 10:00:00',
                   freq='H', tz=pytz.timezone('Etc/GMT+7'))

weather = loc.get_clearsky(dr)
weather['temp_air']  = 25
module_parameters = {'pdc0': 250, 'gamma_pdc': -0.004}
inverter_parameters = {'pdc0': 250}
temperature_parameters = pvlib.temperature.TEMPERATURE_MODEL_PARAMETERS['sapm']['open_rack_glass_glass']
pvs = pvlib.pvsystem.PVSystem(module_parameters=module_parameters,
                              inverter_parameters=inverter_parameters,
                              temperature_model_parameters=temperature_parameters)

mc = pvlib.modelchain.ModelChain(pvs, loc, aoi_model='no_loss',
                                 spectral_model='no_loss',
                                 temperature_model='sapm')
mc.run_model(weather)
mcres = mc.results
print(mcres)


mount = pvlib.pvsystem.FixedMount(surface_tilt=30, surface_azimuth=180,
                                  racking_model='open_rack')
array = pvlib.pvsystem.Array(mount=mount, albedo=0.3,
                             module_parameters=module_parameters,
                             temperature_model_parameters=temperature_parameters)
pvs2 = pvlib.pvsystem.PVSystem(arrays=(array, array),
                             inverter_parameters=inverter_parameters)
mc2 = pvlib.modelchain.ModelChain(pvs2, loc, aoi_model='no_loss',
                                  spectral_model='no_loss',
                                  temperature_model='sapm')
mc2.run_model(weather)
mcres2 = mc2.results
print(mcres2)


mc3 = pvlib.modelchain.ModelChain(pvs2, loc, aoi_model='no_loss',
                                  spectral_model='no_loss',
                                  temperature_model='sapm')
mc3.run_model((weather, weather))
mcres3 = mc3.results
print(mcres3)
```

File: pvlib/modelchain.py [hunk 1]
```diff
@@ -385,6 +412,37 @@ def __setattr__(self, key, value):
...
+        if type(self.dc) is tuple:
+            num_arrays = len(self.dc)
+        else:
+            num_arrays = 1
+
+        desc1 = ('=== ModelChainResult === \n')
+        desc2 = (f'Number of Arrays: {num_arrays} \n')
+        attr = 'times'
+        desc3 = ('Times (first 3)\n' +
+                 f'{_head(_getmcattr(self, attr))}' +
```
Comment (kandersolar) [inline pvlib/modelchain.py:437]:
```suggestion
                 f'{self.times[:3]}' +
```

Seems like this should be fine as long as it's safe to assume `self.times` is a `DatetimeIndex`

File: pvlib/modelchain.py [hunk 2]
```diff
@@ -385,6 +412,37 @@ def __setattr__(self, key, value):
             value = self._result_type(value)
         super().__setattr__(key, value)
 
+    def __repr__(self):
+        # once per MC
+        mc_attrs = ['weather', 'solar_position', 'airmass', 'tracking', 'aoi',
```
Comment (kandersolar) [inline pvlib/modelchain.py:417]: `albedo` is missing

**Review by kandersolar — COMMENTED** [review]
Probably worth a test?  We used to have a `ModelChain.__repr__` test, but it looks like I incorrectly removed it in #1181, oops.  We should probably bring that back.

File: pvlib/modelchain.py [hunk 3]
```diff
@@ -385,6 +412,37 @@ def __setattr__(self, key, value):
...
+        def _head(obj):
+            try:
+                return obj[:3]
+            except:
+                return obj
+
+        if type(self.dc) is tuple:
+            num_arrays = len(self.dc)
+        else:
+            num_arrays = 1
```
[the remark this answers is not in the document]
Comment (cwhanse) [reply pvlib/modelchain.py:431]: This doesn't work since `self` is a `ModelChainResult`, not a `ModelChain`.

File: pvlib/modelchain.py (hunk 1 above)
Comment (cwhanse) [reply pvlib/modelchain.py:433]: Won't work if `times` is a DatetimeStamp. Maybe that can't happen, idk.

**Comment by cwhanse** [comment]
> Probably worth a test? We used to have a `ModelChain.__repr__` test, but it looks like I incorrectly removed it in #1181, oops. We should probably bring that back.

I added the `Modelchain.__repr__` test back.

For the `ModelChainResult.__repr__` test, I'm checking that all the non-private attributes appear. That provides some future-proofing, without a cumbersome string==string check.

File: pvlib/tests/test_modelchain.py [hunk 4]
```diff
@@ -2063,3 +2063,40 @@ def test__irrad_for_celltemp():
     assert len(poa) == 2
     assert_series_equal(poa[0], effect_irrad)
     assert_series_equal(poa[1], effect_irrad)
+
+
+@pytest.mark.parametrize('strategy, strategy_str', [
+    ('south_at_latitude_tilt', 'south_at_latitude_tilt'),
+    (None, 'None')])  # GitHub issue 352
```
Comment (kandersolar) [inline pvlib/tests/test_modelchain.py:2070]: Thanks for putting this test back.  I think this decorator is no longer needed and can be deleted.

File: pvlib/modelchain.py [hunk 5]
```diff
@@ -413,11 +413,7 @@ def __setattr__(self, key, value):
         super().__setattr__(key, value)
 
     def __repr__(self):
-        # once per MC
-        mc_attrs = ['weather', 'solar_position', 'airmass', 'tracking', 'aoi',
-                    'aoi_modifier', 'total_irrad', 'spectral_modifier',
-                    'effective_irradiance', 'cell_temperature', 'diode_params',
-                    'dc', 'dc_ohmic_losses', 'losses', 'ac']
+        mc_attrs = dir(self)
```
Comment (kandersolar) [inline pvlib/modelchain.py:416]: Using `dir` sorts the list of attributes alphabetically instead of conceptually.  Not necessarily a downside.

**Review by kandersolar — APPROVED** [review]

File: pvlib/modelchain.py (hunk 5 above)
Comment (cwhanse) [reply pvlib/modelchain.py:416]: Yeah, once I saw the alphabetized list my thought was "that's easier to use"

**Comment by kandersolar** [comment]
Thanks @cwhanse!

**Pull request merged**